### PR TITLE
Remove Windows specific dependency

### PR DESCRIPTION
**Pull request by Brett208**
For fun, I took a stab at removing the Windows specific code. Unfortunately, I didn't make it.

I was able to check and ensure that the file was a PE file and that it is a DLL. I may have grabbed the right address for the ExportTable, but I doubt it as I didn't know how to translate from RVA to file pointer.

### CODE REVIEW

File: DllExportedVariablesReader32.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#include <array>
+#include <stdexcept>
+#include <cstddef>
+
+struct ImageDataDirectory;
+
+#pragma pack(push, 1) // Make sure structure is byte aligned
+
+struct ImageDataDirectory
+{
```
Comment (DanRStevens) [inline DllExportedVariablesReader32.cpp]: Probably don't need a forward declare here. :wink:

**Review by DanRStevens — COMMENTED** [review]
Looks like a good start.

The following link may help for RVA to file offset conversions:
https://stackoverflow.com/questions/9955744/getting-offset-in-file-from-rva

----

In case it helps, Microsoft has some PE format documentation:
https://docs.microsoft.com/en-us/windows/win32/debug/pe-format

----

I find it fun that you're going down this road. I'm sure it will be educational.

**Comment by Brett208** [comment]
The code isn't well written, but I think it is cross platform compatible now. I managed to read all the exported variables from the PE format without loading the dll into memory. The Microsoft documentation on the PE Format is difficult to read.

I also suspect the code is now compatible for compilation in x86 or x64. I'll try this out later.

@DanRStevens I'd be curious if you could bring it up on Linux and see what errors and warnings are reported.

Thanks,
Brett

File: DllExportedVariablesReader32.cpp [hunk 2]
```diff
@@ -1,58 +1,119 @@
...
+// File Offset = RVA - Virtual Offset + Raw Offset.
+std::uint32_t DllExportedVariableReader32::RvaToFileOffset(std::uint32_t rva, const SectionTable& sectionTable)
+{
+	return rva - sectionTable.virtualAddress + sectionTable.pointerToRawData;
 }
 
-bool DllExportedVariableReader32::IsPortableExecutableFile(Stream::FileReader& stream)
+bool DllExportedVariableReader32::IsPortableExecutableFile()
 {
 	stream.SeekForward(0x3c);
```
Comment (DanRStevens) [inline DllExportedVariablesReader32.cpp]: Seems like this would be more robust as an absolute seek.

**Review by DanRStevens — APPROVED** [review]
Looks good. I'm curious, does it run any faster now?

----

On Linux, I'm still getting failures from all compilers:

Mingw:
`#include <filesystem>` fails. Changing to `<experimental/filesystem>` works, but code still fails at the link step due to undefined references to `fs::path::generic_string()`. I did include `-lstdc++fs`, and at the end of the command line, so that's not the problem. Example command:
```
i686-w64-mingw32-g++ -IOP2Utility/include/ -IOutpost2DLL/ -std=c++17 *.cpp -LOP2Utility/ -lOP2Utility -lstdc++fs
```

Clang:
Fails on `#include <windows.h>` in MissionScanner.cpp. Fails on `OP2` macro expansion with MSVC specific `__declspec`. Adding `-DOP2=""` works around the macro expansion problem. Fails at link step due to incompatible `libOP2Utility.a`. Example command:
```
clang -DOP2="" -IOP2Utility/include/ -IOutpost2DLL/ -std=c++17 *.cpp -LOP2Utility/ -lOP2Utility -lstdc++fs
```

GCC:
Fails on `#include <windows.h>`. Fails on `OP2` macro expansion. Fails on `__fastcall` and `__cdecl` specifiers. Fails at link step due to incompatible `libOP2Utility.a`. Example command:
```
g++ -DOP2="" -D__fastcall="" -D__cdecl="" -IOP2Utility/include/ -IOutpost2DLL/ -std=c++17 *.cpp -LOP2Utility/ -lOP2Utility -lstdc++fs
```

**Comment by DanRStevens** [comment]
After recompiling OP2Utility with Clang, I was able to build this project with Clang. The result runs, and displays a summary table. Compile command:
```
clang++ -DOP2="" -D__fastcall="" -IOP2Utility/include/ -IOutpost2DLL/ -std=c++17 *.cpp -LOP2Utility/ -lOP2Utility -lstdc++fs
```

Similarly, after recompiling OP2Utility with GCC, I was able to build this project with GCC. The result runs as expected. Compile command:
```
g++ -DOP2="" -D__fastcall="" -D__cdecl="" -IOP2Utility/include/ -IOutpost2DLL/ -std=c++17 *.cpp -LOP2Utility/ -lOP2Utility -lstdc++fs
```

----

Edit:
Nicely done with the PE reading code. :)

File: DllExportedVariablesReader32.cpp (hunk 2 above)
Comment (Brett208) [reply DllExportedVariablesReader32.cpp]: Yeah, good point.

**Comment by Brett208** [comment]
Thanks for the quick checkout on Linux. I'm glad it wasn't too tough to get working.

Since you asked, I ran the numbers using the Timer ~~I found on the internet~~ developed from scratch: units are seconds
Using LoadLibrary
1.39642
1.41721
1.35062
AVG: 1.388

Using stream
0.891229
1.04335
0.960045
AVG: .965

So, not loading it into memory like a proper dll was roughly 40% faster. 

The stopwatch was placed at the beginning and end of the main function. Assuming the actual writing of the table (cout calls) is the same between both methods, traversing the dll without loading it into memory is significantly faster. Especially considering I did little to no optimization of the stream code. Interesting.

Learning a little about the PE format helps a lot with understanding how exes and dlls are loaded into memory and talk to each other. By no means could I be considered knowledgeable about the subjects.

My goal was making it able to compile on Linux and learn the PE format. I suppose it is an improvement and worth merging to remove Windows requirement. Did you want to spend time reviewing this code in the near future? Otherwise I could just merge it now. It wasn't written with neat commits, sorry about that. I was more worried about the exploratory learning.

Thanks,
Brett

**Comment by DanRStevens** [comment]
Hmm, that's a nice little speedup.

I think it's worth merging. The code looked fine on first inspection. The commits don't need to be perfectly clean, and we can always update the code later if you think something needs improvement.

Glad you learned something. I find the PE format is quite interesting.

----

As a bit of a side thing, while the PE format is fresh in your head, I might suggest taking a look at the relocation section of the PE format. Access to global variables uses fixed addresses, but if the executable is relocated, those fixed addresses need to be updated. Hence the relocation section. By traversing that data, you can find all accesses to global variables. Very neat. Arklon used that trick in some of his patch code. As we further develop the patching in op2ext, and increase the automation of patching, we will likely want to make use of that trick.

**Comment by Brett208** [comment]
Thanks

**Comment by DanRStevens** [comment]
Tagging issue #5 as related.

**Pull request merged**